framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -1210,4 +1212,28 @@ private void addAnnotationToStore(CFStore store, AnnotationMirror anno, Node nod
                 resultValues,
                 transferResult.getResultValue().getUnderlyingType());
     }
+
+    /**
+     * The value checker doesn't override {@link
+     * org.checkerframework.framework.flow.CFAbstractTransfer#strengthenAnnotationOfEqualTo(TransferResult,
+     * Node, Node, CFAbstractValue, CFAbstractValue, boolean)}, which is called by {@link
+     * org.checkerframework.framework.flow.CFAbstractTransfer#visitCase(CaseNode, TransferInput)}.
+     * Therefore, we need to explicitly call the handling for binary comparison if we want to do the
+     * extra refinements that the Value Checker applies on switch statements.
+     */
+    @Override
+    public TransferResult<CFValue, CFStore> visitCase(
+            CaseNode n, TransferInput<CFValue, CFStore> p) {
+        TransferResult<CFValue, CFStore> transferResult = super.visitCase(n, p);
+        if (n.getSwitchOperand() instanceof FieldAccessNode) {
```

Comment:
This problem only came up when trying to refine `array.length` field accesses, but it is more general, so the `if` guard can (and should) probably be removed.